### PR TITLE
enh(logs): enhance php sql failures logging

### PR DESCRIPTION
**Pull request by Nohzoh**
## Description

Log all SQL error messages.

Also some refacto:

* remove useless parameters
* better variables naming
* and other code factorization/simplification

## How this pull request can be tested ?

Either there are known behavior logging this kind of error.
Or some piece of code could be modified to cause such errors.
As NR, there are many things makes SQL request that this could be validated by many ways (exemple: fresh install + configuration + monitoring should be interesting to test).

### CODE REVIEW

File: www/class/centreonDB.class.php [hunk 1]
```diff
@@ -86,7 +85,7 @@ class CentreonDB extends \PDO
      *
      * @throws Exception
      */
-    public function __construct($db = self::LABEL_DB_CONFIGURATION, $retry = 3, $silent = false)
+    public function __construct($db = self::LABEL_DB_CONFIGURATION, $retry = 3)
```
Comment (kduret) [inline www/class/centreonDB.class.php:88]: `$silent` must be removed from phpdoc

File: www/class/centreonDBStatement.class.php [hunk 2]
```diff
@@ -44,36 +44,30 @@
...
+    protected function __construct(CentreonLog $log = null)
     {
-        $this->dbh = $dbh;
         $this->log = $log;
-        $this->debug = $debug ? true : false;
-        $this->fetchAll = null;
+        $this->allFetched = null;
     }
 
     public function fetch($fetch_style = null, $cursor_orientation = PDO::FETCH_ORI_NEXT, $cursor_offset = 0)
```
Comment (AssanDiomande) [inline www/class/centreonDBStatement.class.php:63]: Missing phpDoc

File: www/class/centreonDBStatement.class.php [hunk 3]
```diff
@@ -89,23 +83,23 @@ public function free()
 
     public function numRows()
```
Comment (AssanDiomande) [inline www/class/centreonDBStatement.class.php:84]: Missing phpdoc

File: www/class/centreonDBStatement.class.php [hunk 4]
```diff
@@ -89,23 +83,23 @@ public function free()
...
-        if (is_null($this->fetchAll)) {
-            $this->fetchAll = $this->fetchAll();
+        if (is_null($this->allFetched)) {
+            $this->allFetched = $this->fetchAll();
         }
-        return count($this->fetchAll);
+        return count($this->allFetched);
     }
 
     public function execute($parameters = null)
```
Comment (AssanDiomande) [inline www/class/centreonDBStatement.class.php:92]: Missing phpDoc

**Review by AssanDiomande — DISMISSED** [review]

File: www/class/centreonDB.class.php [hunk 5]
```diff
@@ -48,23 +48,75 @@
...
+     */
     public const LABEL_DB_CONFIGURATION = 'centreon';
+
+    /**
+     * @var string
+     */
     public const LABEL_DB_REALTIME = 'centstorage';
+
+    /**
+     * @var array<mixed>
```
Comment (kduret) [inline www/class/centreonDB.class.php:62]:
```suggestion
     * @var array<string,\CentreonDB>
```

File: www/class/centreonDB.class.php [hunk 6]
```diff
@@ -48,23 +48,75 @@
...
+     */
     protected $db_port = "3306";
+
+    /**
+     * @var int
+     */
     protected $retry;
+
+    /**
+     * @var array<mixed>
```
Comment (kduret) [inline www/class/centreonDB.class.php:82]:
```suggestion
     * @var array<string,mixed>
```

File: www/class/centreonDB.class.php [hunk 7]
```diff
@@ -48,23 +48,75 @@
...
+     */
     protected $dsn;
+
+    /**
+     * @var (int|(string|\CentreonLog[])[]|string|true)[]
+     */
     protected $options;
+
+    /**
+     * @var mixed
```
Comment (kduret) [inline www/class/centreonDB.class.php:92]:
```suggestion
     * @var string
```

File: www/class/centreonDB.class.php [hunk 8]
```diff
@@ -232,21 +276,22 @@ protected function displayConnectionErrorPage($msg = null)
...
-        return $escapedStr;
+        return addslashes($str);
     }
 
     /**
      * Query
      *
-     * @return CentreonDBStatement
+     * @return PDOStatement|null
+     * @param mixed $queryString
```
Comment (kduret) [inline www/class/centreonDB.class.php:292]:
```suggestion
     * @param string $queryString
```

File: www/class/centreonDB.class.php [hunk 9]
```diff
@@ -286,10 +331,11 @@ public function query($queryString, $parameters = null, ...$parametersArgs)
      * launch a getAll
      *
      * @access public
-     *
+     * @throws \PDOException
      * @param string $query_string query
+     * @param array<mixed> $placeHolders
      *
-     * @return  object  getAll result
+     * @return array<mixed>|false  getAll result
```
Comment (kduret) [inline www/class/centreonDB.class.php:338]:
```suggestion
     * @return mixed[]|false  getAll result
```

File: www/class/centreonDB.class.php [hunk 10]
```diff
@@ -346,8 +390,9 @@ public function numberRows()
 
     /**
      * checks if there is malicious injection
+     * @param mixed $sString
```
Comment (kduret) [inline www/class/centreonDB.class.php:393]:
```suggestion
     * @param string $sString
```

File: www/class/centreonDBStatement.class.php [hunk 11]
```diff
@@ -44,68 +44,94 @@
...
-        $this->debug = $debug ? true : false;
-        $this->fetchAll = null;
+        $this->allFetched = null;
     }
 
+    /**
+     * This method overloads PDO `fetch` in order to use the possible already
+     * loaded data available in `allFetched`.
+     *
+     * @inheritDoc
```
Comment (kduret) [inline www/class/centreonDBStatement.class.php:75]:
```suggestion
     * {@inheritDoc}
```

File: www/class/centreonDBStatement.class.php [hunk 12]
```diff
@@ -44,68 +44,94 @@
...
+            $this->allFetched = $this->fetchAll();
         }
-        return count($this->fetchAll);
+        return count($this->allFetched);
     }
 
-    public function execute($parameters = null)
+    /**
+     * This method wraps the PDO `execute` method and manages failures logging
+     * @inheritDoc
```
Comment (kduret) [inline www/class/centreonDBStatement.class.php:122]:
```suggestion
     * {@inheritDoc}
```

**Review by kduret — APPROVED** [review]

**Review by kduret — CHANGES_REQUESTED** [review]
error when running unit tests : 
`PHP Fatal error:  Declaration of Centreon\Test\Mock\CentreonDB::autoCommit($enable) must be compatible with CentreonDB::autoCommit($val): void in /usr/local/src/centreon-web/vendor/centreon/centreon-test-lib/src/mock/CentreonDB.php on line 204`

**Comment by AssanDiomande** [comment]
> error when running unit tests : `PHP Fatal error: Declaration of Centreon\Test\Mock\CentreonDB::autoCommit($enable) must be compatible with CentreonDB::autoCommit($val): void in /usr/local/src/centreon-web/vendor/centreon/centreon-test-lib/src/mock/CentreonDB.php on line 204`

This should be ok now, I reverted phpdoc of autocommit function to match the one ine vendor

File: www/class/centreonDB.class.php [hunk 13]
```diff
@@ -160,25 +202,29 @@ public function __construct($db = self::LABEL_DB_CONFIGURATION, $retry = 3, $sil
...
                 $this->displayConnectionErrorPage($e->getMessage());
             } else {
                 throw new Exception($e->getMessage());
             }
         }
     }
 
+    /**
+     * @param type $enable
+     * @return type
```
Comment (kduret) [inline www/class/centreonDB.class.php:215]:
```suggestion
     * @param mixed $val
     * @return void
```

**Review by kduret — CHANGES_REQUESTED** [review]

**Review by kduret — DISMISSED** [review]

**Review by AssanDiomande — DISMISSED** [review]

File: bin/centreon-partitioning.php [hunk 14]
```diff
@@ -51,11 +51,17 @@
...
 $partEngine = new PartEngine();
 
 if (!$partEngine->isCompatible($centstorageDb)) {
-    exitProcess(PROCESS_ID, 1, "[".date(DATE_RFC822)."] CRITICAL: MySQL server is not compatible with partitioning. MySQL version must be greater or equal to 5.1\n");
+    exitProcess(
+        PROCESS_ID,
+        1,
+        "[" .
+        date(DATE_RFC822) .
+         "] CRITICAL: MySQL server is not compatible with partitioning. MySQL version must be greater or equal to 5.1\n"
```
Comment (kduret) [inline bin/centreon-partitioning.php:63]:
```suggestion
        "[" . date(DATE_RFC822) . "] "
        . "CRITICAL: MySQL server is not compatible with partitioning. MySQL version must be greater or equal to 5.1\n"
```

File: cron/centreon-partitioning.php [hunk 15]
```diff
@@ -45,19 +45,25 @@
...
 $partEngine = new PartEngine();
 
 if (!$partEngine->isCompatible($centstorageDb)) {
-    exitProcess(PROCESS_ID, 1, "[".date(DATE_RFC822)."] CRITICAL: MySQL server is not compatible with partitionning. MySQL version must be greater or equal to 5.1\n");
+    exitProcess(
+        PROCESS_ID,
+        1,
+        "[" .
+        date(DATE_RFC822) .
+        "] CRITICAL: MySQL server is not compatible with partitionning. MySQL version must be greater or equal to 5.1\n"
```
Comment (kduret) [inline cron/centreon-partitioning.php:57]:
```suggestion
        "[" . date(DATE_RFC822) . "] "
        . "CRITICAL: MySQL server is not compatible with partitionning. MySQL version must be greater or equal to 5.1\n"
```

File: src/Centreon/Application/Webservice/AclGroupWebservice.php [hunk 16]
```diff
@@ -137,7 +143,6 @@ public static function dependencies(): array
      * Get a list of ACL groups
      *
      * @throws \RestBadRequestException
```
Comment (kduret) [inline src/Centreon/Application/Webservice/AclGroupWebservice.php:145]:
```suggestion
     * @return DataRepresenter\Response
     * @throws \RestBadRequestException
```

File: src/Centreon/Application/Webservice/AclGroupWebservice.php [hunk 17]
```diff
@@ -45,8 +50,9 @@ public static function getName(): string
 
     /**
      * {@inheritdoc}
+     * @return array<int,string>
      */
-    public static function dependencies(): array
+    public static function dependencies()
```
Comment (kduret) [inline src/Centreon/Application/Webservice/AclGroupWebservice.php:55]:
```suggestion
    public static function dependencies(): array
```

File: src/Centreon/Domain/Repository/TopologyRepository.php [hunk 18]
```diff
@@ -75,6 +75,8 @@ public function enableMenus(): bool
     /**
      * Get Topologies according to ACL for user
      * @todo refactor this into function below it
+     * @return array<mixed>
+     * @param mixed $user
```
Comment (kduret) [inline src/Centreon/Domain/Repository/TopologyRepository.php:79]:
```suggestion
     * @param mixed $user
     * @return mixed[]
```

File: src/Centreon/Domain/Repository/TopologyRepository.php [hunk 19]
```diff
@@ -185,13 +187,13 @@ public function getTopologyList(CentreonUser $user): array
     /**
      * Find Topology entity by criteria
      *
-     * @param array $params
+     * @param array<mixed> $params
```
Comment (kduret) [inline src/Centreon/Domain/Repository/TopologyRepository.php:190]:
```suggestion
     * @param mixed[] $params
```

**Review by kduret — DISMISSED** [review]

File: src/Centreon/Domain/Repository/TopologyRepository.php [hunk 20]
```diff
@@ -151,7 +153,7 @@ public function getReactTopologiesPerUserWithAcl($user)
     /**
      * Get list of topologies per user and filter by react pages if specified
      * @param CentreonUser $user
-     * @return array
+     * @return array<mixed>|false
      */
     public function getTopologyList(CentreonUser $user): array
```
Comment (kduret) [inline src/Centreon/Domain/Repository/TopologyRepository.php:158]:
```suggestion
    public function getTopologyList(CentreonUser $user)
```

File: features/bootstrap/ClapiContext.php [hunk 21]
```diff
@@ -18,7 +18,6 @@ class ClapiContext extends CentreonContext
         "TRAP",
         "HTPL",
         "CG",
-        "LDAP",
```
Comment (kduret) [inline features/bootstrap/ClapiContext.php:21]: do not delete this and checkout `MON-12225-enhance-php-sql-failures-logging` on centreon-build from `develop`

Comment (AssanDiomande) [reply features/bootstrap/ClapiContext.php]: done

File: features/Clapi.feature [hunk 22]
```diff
@@ -7,7 +7,7 @@ Feature: Clapi
     Given a Centreon server
     When the user uses the clapi export command
     Then a valid clapi configuration file should be generated
-    And it should contain the supported configuration objects
+    Then it should contain the supported configuration objects
```
[the remark this answers is not in the document]
Comment (AssanDiomande) [reply features/Clapi.feature]: done

File: www/class/centreonDB.class.php [hunk 23]
```diff
@@ -48,23 +48,75 @@
...
+     */
     protected $retry;
+
+    /**
+     * @var array<string,mixed>
+     */
     protected $dsn;
+
+    /**
+     * @var (int|(string|\CentreonLog[])[]|string|true)[]
```
Comment (callapa) [inline www/class/centreonDB.class.php:87]: ???

File: www/class/centreonDB.class.php [hunk 24]
```diff
@@ -160,25 +202,29 @@ public function __construct($db = self::LABEL_DB_CONFIGURATION, $retry = 3, $sil
...
+            if (php_sapi_name() !== "cli") {
                 $this->displayConnectionErrorPage($e->getMessage());
             } else {
                 throw new Exception($e->getMessage());
             }
         }
     }
 
+    /**
+     * @param mixed $enable
```
Comment (callapa) [inline www/class/centreonDB.class.php:214]: $enable ???

Comment (AssanDiomande) [reply www/class/centreonDB.class.php:214]: updated

File: www/class/centreonDB.class.php (hunk 23 above)
Comment (AssanDiomande) [reply www/class/centreonDB.class.php:87]: warning message from phpstan

Comment (AssanDiomande) [reply www/class/centreonDB.class.php]: updated phpdoc

File: www/class/centreonDB.class.php [hunk 25]
```diff
@@ -160,25 +202,29 @@ public function __construct($db = self::LABEL_DB_CONFIGURATION, $retry = 3, $sil
...
     }
 
+    /**
+     * @param mixed $val
+     * @return void
+     */
     public function autoCommit($val)
     {
         /* Deprecated */
     }
```
Comment (jeremyjaouen) [inline www/class/centreonDB.class.php:220]: I think it could be remove has this function just do nothing.

**Review by jeremyjaouen — APPROVED** [review]

File: www/class/centreonDB.class.php (hunk 25 above)
Comment (jeremyjaouen) [reply www/class/centreonDB.class.php:220]: We'll let it like this to prevent side effect

**Review by kduret — APPROVED** [review]

**Pull request merged**